### PR TITLE
Search bar feature.

### PR DESCRIPTION
**Pull request by iCHAIT**
Right now this PR is not in a position to be merged.
Currently I have implemented the search bar feature for the vendors form only and before I implement it for products,preparations,images,stories,videos as well I thought it would be better if I had a discussion here.
I have a few things in mind that I wanted to get cleared -
1.Is the search bar form that I have provided fine and acceptable or are there any changes that you want me to do?
2.Consider a case, suppose the name of the vendor is "Alice" now if the user searches for "alice" then what should the result be??
Should it return Alice only or should the search yield no results.
In short how should we deal with case sensitivity?

Anything else that you guys would like to mention?

### CODE REVIEW

**Comment by tschuy** [comment]
The way you've implemented search is very rigid. It doesn't do partial matches, lowercases, etc. 

Rather than rolling our own search with a series of complicated filters, the best way to implement this would probably be using something pre-built like [Haystack](http://haystacksearch.org/) with the Whoosh backend (other projects at the OSL use this setup, though I've not had the opportunity to look at how they implement search other than knowing they use Haystack and Whoosh). This would allow the user to search for vendors, products, etc all at the same time, and we could allow them to filter by the actual type as an addtional option.

**Comment by iCHAIT** [comment]
Thanks for pointing out the shortcomings of the current approach, I'll fix this using Haystack and update this PR.

**Comment by iCHAIT** [comment]
@tschuy I have implemented the search bar feature ( currently working for the vendors form only) using Haystack with the whoosh backend, currently Travis fails and it is inevitable because the code I've written  handles vendors and not any other form ( so it is bound to fail for other entities as is the case right now), so please ignore it as of now and kindly review the code.

Before proceeding further and implementing the same for other forms as well I want to know whether I'm on the right path?

**Comment by tschuy** [comment]
Very impressive! Sorry for not getting back to you sooner. 

I've currently broken my development environment, so I'm going to need to wait until tomorrow to actually run your code, but it looks good. 

For the travis tests, the reason they appear to be failing is because you've changed the iterables being output to the form from item to item.object. Again, I'll take a closer look tomorrow when I have time to reset my dev environment.

**Comment by tschuy** [comment]
Ok, I just finished playing with the new search feature. It works just as expected!

Looking over the tests, you're exactly right about the tests, though it does look like the ListVendorTestCase is failing too, since it's expecting a QuerySet and you're now passing it a SearchResult. It also looks like your SearchResult doesn't default to ordering by name as the QuerySet does. A quick ``.order_by('name')`` tacked on when search is None should take care of this.

You've not written any new tests for the searches. This shouldn't be too difficult; you could just pick a search result for which there should only be one or two results, and then test that they show up when you search. You might need to set up the connection in each test that tests the searches, though. This blog post looks helpful:

http://bwreilly.github.io/blog/2013/07/21/testing-search-haystack-in-django/

These tests can just be added to the existing List*TestCase tests.

**Comment by tschuy** [comment]
http://django-haystack.readthedocs.org/en/latest/searchquery_api.html#sq-objects <-- you'll want to sanitize user input.

**Comment by tschuy** [comment]
For some reason, I can't access the object by calling ``.object`` on the objects. I'm not sure why...

File: whats_fresh/whats_fresh_api/templates/list.html [hunk 1]
```diff
@@ -53,3 +62,7 @@ <h2 class='light'>{{ title }}</h2>
 </div>
     </div>
 {% endblock content %}
+
+<script>
+$('#search_list').attr("value", "kokab");
```
Comment (tschuy) [inline whats_fresh/whats_fresh_api/templates/list.html]: What is this for?

Comment (iCHAIT) [inline whats_fresh/whats_fresh_api/templates/list.html]: Oops, this was when I was trying to retain the value of searches!
Will remove it :p

Comment (iCHAIT) [inline whats_fresh/whats_fresh_api/templates/list.html]: I thought I'll need jquery for that.

File: whats_fresh/whats_fresh_api/templates/list.html [hunk 2]
```diff
@@ -9,14 +9,22 @@
...
 {% block content %}
+<form class="search-bar" action="">
+<input id="search-list" type="text" name="search" {% if search_text %} value="{{ search_text }}" {% endif %} required="true">
+<button id="search-btn" type="submit">Search</button>
+</form><br>
 <div class='message'><h4 class='light accent'>{{ message }}</h4></div>
     <div class='list'>
 {% for item in item_list %}
+{% if item != None %}
+{% comment %}
```
Comment (tschuy) [inline whats_fresh/whats_fresh_api/templates/list.html]: I would make this comment all on one line, and clarify that with Whoosh, sometimes items `in` items are None -- perhaps "Bad workaround for Whoosh searches containing None items"?

File: whats_fresh/whats_fresh_api/views/entry/images.py [hunk 3]
```diff
@@ -31,7 +33,21 @@ def image_list(request):
     elif request.GET.get('saved') == 'true':
         message = "Entry saved successfully!"
 
-    paginator = Paginator(Image.objects.order_by('name'), settings.PAGE_LENGTH)
+    if request.GET.get('search') is None or request.GET.get('search') == "":
+        images = Image.objects.order_by('name')
+    else:
+        if request.GET.get('search') != "":
+            images = list(OrderedDict
```
Comment (tschuy) [inline whats_fresh/whats_fresh_api/views/entry/images.py]: these indentations are getting a little silly :P Is there a way to do this that doesn't have the content_auto indented 50 characters?

**Comment by tschuy** [comment]
Other than the indentation and that comment, I think it's good to go once you get those tests working (which is an issue with not doing the OrderedDict part, I believe?).

File: whats_fresh/whats_fresh_api/views/entry/stories.py [hunk 4]
```diff
@@ -28,7 +30,19 @@ def story_list(request):
...
-    paginator = Paginator(Story.objects.order_by('name'), settings.PAGE_LENGTH)
+    if request.GET.get('search') is None or request.GET.get('search') == "":
+        stories = Story.objects.order_by('name')
+    else:
+        if request.GET.get('search') != "":
+            stories = list(
+                OrderedDict.fromkeys(
+                    item.object for item in
+                    SearchQuerySet().models(Story).autocomplete(
+                        content_auto=request.GET.get('search', ''))))
```
Comment (iCHAIT) [inline whats_fresh/whats_fresh_api/views/entry/stories.py]: @tschuy Does this look better?

**Comment by iCHAIT** [comment]
@tschuy  I think we should get rid of these - 

https://github.com/osu-cass/whats-fresh-api/blob/develop/whats_fresh/whats_fresh_api/functions.py#L8

and 

https://github.com/osu-cass/whats-fresh-api/blob/develop/whats_fresh/whats_fresh_api/models.py#L291

As they don't solve the purpose we included them for.

What do you reckon?

**Comment by tschuy** [comment]
Looks good to me.
